### PR TITLE
Added LU factorization, determinant and direct solve procedures

### PR DESCRIPTION
**Pull request by wongalvis14**
Add LU factorization, matrix determinant calculation and exact solve (with LU) procedures to the dense LinearAlgebra library, all written in Chapel only

### CODE REVIEW

**Comment by wongalvis14** [comment]
Although it is a drawback that only dense L,U would be returned, it gives an advantage to build determinant, inverse, etc. routines easily for sparse matrices in conjunction with LU factorization and other routines in Chapel. On a side note, sparse matrix inverse does not necessarily end up with a sparse matrix.

**Comment by wongalvis14** [comment]
To solve a linear system

```
// A * x = b, A and b are pre-defined
var (L,U) = lu(A);
var z = solve_tril(L, b);
var x = solve_triu(U, z);
```

**Comment by e-kayrakli** [comment]
Nothing obviously wrong at my first look except for the correctness test: could you make it a proper correctness test by making sure that input is reasonably small with `--correctness` flag and it is printed out. (and `.good` file has that output)

I will give this a closer look soon.

**Comment by npadmana** [comment]
@wongalvis14, @e-kayrakli -- thanks for all your work putting all of this functionality into LinearAlgebra. Unfortunately, this particular algorithm isn't numerically stable; that requires a pivoting algorithm.

Here's a failing example (run off your branch) -- 
```chapel
use LinearAlgebra;

var A = Matrix([0.0,1.0],
               [1.0,2.0]);
writeln("A=",A);
var (l,u) = lu(A);
writeln("L=",l);
writeln("U=",u);
```
which yields 
```
A=0.0 1.0
1.0 2.0
L=1.0 0.0
inf 1.0
U=0.0 1.0
0.0 -inf
```

**Comment by wongalvis14** [comment]
Thank you. I misunderstood that there's an algorithm that can factorize without any pivoting and does not fail. I will add pivoting, the determinant algorithm will need to be modified, and the two trilu solve proc's don't need to be changed I think.

**Comment by npadmana** [comment]
You may need to pass the permutations to the solve procs here. 

Not to be too greedy, but it would also be great if this routine defaulted to the LAPACK routines for local, dense matrices, just as you did with the inverse case.

**Comment by wongalvis14** [comment]
If we have a `solve()` wrapper, I think we'll take the permutation as an argument. Otherwise, it might be better to leave multiplying `Pb` to the user.

**Comment by wongalvis14** [comment]
@npadmana  I've added pivoting and it worked fine on the counter-example you gave, please help me test a bit more with some weird edge cases. Thanks a ton!!! 

Also, I'm figuring out a neat and fast way to calculate the determinant of the permutation matrix. Do you have an idea on that too?

**Comment by npadmana** [comment]
@wongalvis14 --- on the determinant of a permutation matrix, it is (-1)^(# of transpositions). If I think a permutation as a product of cycles, the determinant is \prod_{cycles} (-1)^{length of cycle-1}. And I think you can work out the length of a cycle by just walking the permutation. Although that might not be the easiest thing to parallelize. So see below.

A different way to do this is to count the number of inversions (i.e. when i < j, \pi(i) > \pi(j)). That's easy to parallelize, but is O(N^2) naively. However, a little googling leads to :
https://stackoverflow.com/questions/6523712/calculating-the-number-of-inversions-in-a-permutation

which is O(N log N), using a merge sort. The determinant is then (-1)^(number of inversions). I think this would be the way to do this. 

I might suggest storing the permutation matrix simply as a list of N integers (for an NxN matrix). Eg. for a 4x4 matrix, you might have 2 1 3 4 which swaps the first and second rows. 

I'll also add that while doing LU, you already know the number of transpositions/swaps. So if you dump that out as well, the determinant requires zero work.

I'll think about test cases and send those along next.

**Comment by npadmana** [comment]
Also, agreed on `triu` and `tril` not doing `P` and leaving that to `solve`.

**Comment by npadmana** [comment]
One last comment -- as I've been looking at this, I found 
https://epubs.siam.org/doi/book/10.1137/1.9780898718027?mobileUi=0
useful.

**Comment by e-kayrakli** [comment]
Thanks for the comments @npadmana !

@wongalvis14  -- If we are going in the direction of using a merge-sort like algorithm to count the inversions, a parallel merge sort is implemented in the `Sort` package already. Not that it can be used as is for this purpose, but there may be significant amount of code reuse. See the [MergeSort module](https://github.com/chapel-lang/chapel/blob/master/modules/packages/Sort.chpl#L881)

**Comment by wongalvis14** [comment]
@npadmana @e-kayrakli Thanks for the feedback. I think I'll overload a function or write a private proc that reuses the code that returns the no. of swaps directly. It'll definitely be the most efficient way.

**Comment by e-kayrakli** [comment]
Also, the CI fails b/c you have used tabs in your code. You should use spaces instead. But it doesn't mean you have to push space multiple times. Look up [soft tabs](https://stackoverflow.com/questions/26350689/what-are-hard-and-soft-tabs)

**Comment by npadmana** [comment]
@wongalvis14  -- linked is a gist with a set of tests for the new LU code. A few of these do fail, but that is by design; I've chosen matrices that are known to cause trouble with a partial pivoting strategy.

And many do pass, so nice work!

I don't think it's immediately necessary to fix these, but it would be interesting to see what LAPACK does in the cases that fail (I believe the default LAPACK strategy is partial pivoting, but I wonder if it does some error checks to see if this is somehow failing). If these both silently fail, then I think we can document this issue, but proceed otherwise (I believe LAPACK also has a full pivoting option, but I don't think that is the default). Otherwise, we'll need to dig some more.

https://gist.github.com/npadmana/5e8c7e15ac4d142e8c1eb2d2b39ff993

**Comment by wongalvis14** [comment]
@npadmana Wow! 

I didn't move on to full pivoting because I think that routine may take up too much memory and the subsequent determinant / solve routines will be too inefficient. 

And for fail are you referring to inaccurate answers, instead of halt/crashes?

Last but not least, can we adopt your script as the correctness test?

**Comment by npadmana** [comment]
Yes, the answers appear to be wrong; the code does not crash. I'm curious what happens with LAPACK.

I don't think full pivoting should be more expensive in terms of memory if you store the permutation matrices as a simple N-element array. 

Feel free to use any/all of the code as a correctness test. I would also suggest generating  matrices along the lines of the Matlab matrix test gallery as some further test cases (some of the ones I used are here as well -- I got them from the references here).

https://www.mathworks.com/help/matlab/ref/gallery.html

**Comment by wongalvis14** [comment]
For users like me (or us) of course storing the permutations as a N-element array is preferred, but others may prefer having it as a matrix for convention. That said I'm not sure if there are some popular routines in scientific or numerical computing that needs the permutation matrix in a full matrix form in order to easily feed it into other functions.
And MATLAB also returns a full array iirc, but it's MATLAB.

**Comment by npadmana** [comment]
Re partial pivoting -- it looks like the standard routines all just implement partial pivoting, and rely on the user to know/check if LU factorization worked or not. And it's also is true that the simple partial pivoting works in a very large number of cases. So I think it's fine for this routine to just implement partial pivoting (at least for now, as the default). And you can safely ignore the failures in the tests I sent.

On the permutation matrix, it looks like MATLAB has two options -- one to return as a vector, and the other to return as a matrix. Every other routine I've seen just has it returned as a vector. Multiplying by a permutation matrix is much simpler than actually doing the matrix multiplication. And in most cases, the user just passes this from one to the other. Most of the time, you rarely care about the LU decomposition -- you just use it to do something, like solve.

Which does bring up the interesting question -- why don't we repackage the entire LU decomposition in a class? It would be trivial then to keep auxiliary information (like the number of row interchanges) and we can build methods like `solve` for it.

And anyone can get at all the various pieces easily enough. In fact, we can define methods that would give either a `Matrix` view onto objects like `P` or just construct them if a user really wanted it.

And it could easily be extensible. One might imagine a `LU` class with `PartialLU` and `FullLU` as a subclass. And `LU` could be part of a more general `MatrixFactorization` class at some point.

Finally, I might suggest packing both `L` and `U` into the same matrix.....

**Comment by wongalvis14** [comment]
Yes, packing `L` and `U` into the same matrix is a good idea, I'm sorry I overlooked that in MATLAB there's a `Y = lu(X)` way to call `lu` which does what you suggested. And I'll follow the `dgetrf()` format to return the permutation matrix.

**Comment by e-kayrakli** [comment]
I think packaging LU in its own class is a very good idea

However, my immediate reaction is that it should be part of a longer term effort where the LinearAlgebra module makes use of more object-oriented approach.  So that having a class family for LU would feel more organic. But that's just my opinion. There may be performance caveats on that route, as well.

Also, I (more strongly) think that if it is important to get that interface sooner rather than later, it should come in as a separate PR where the design here is refactored into a class.

**Comment by wongalvis14** [comment]
Example of solve now:
```
var B: [{1..3,1..3}] real;
B[1,..] = [1,0,-5];
B[2,..] = [2,6,-2];
B[3,..] = [2,5,2];
var b: [{1..3}] real = [1,2,3];

var (LUb, ipivb) = lu(B);
b = permute(ipivb, b, true);

var z = solve_tril(LUb, b);
var x = solve_triu(LUb, z);
```

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+      forall j in 1..(i-1) by -1 {
+        y(j) -= U(j,i) * sol;
+      }
+    }
+  }
+  
+  return y;
+}
+
+/* Returns the solution ``x`` to the linear system `` A * x = b ``.
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Apparently sphinx doesn't enjoy seeing `<space>``.` Removing spaces around `A*X=B` or removing the full period makes the doc compilable. It doesn't seem like the existing docs are very consistent, so I don't have a strong preference. But while there you should change the tense from present to imperative. i.e. not `Returns` but `Return`. Similarly, you should make that change for other functions, as well.

**Comment by wongalvis14** [comment]
@e-kayrakli @ben-albrecht Ready for review

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
   return trace;
 }
 
+proc _lu (in A: [?Adom] ?eltType) {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: This should be a `private proc`

File: modules/packages/LinearAlgebra.chpl [hunk 3]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+  const n = Adom.shape(1);
+  const LUDom = {1..n, 1..n};
+
+  // TODO: Reduce memory usage
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Just to be clear, you are talking about wasted memory for storing triangular matrices as full matrices, right?

File: modules/packages/LinearAlgebra.chpl [hunk 4]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+/*
+  Compute an LU factorization of square matrix `A` 
+  using partial pivoting, such that `A = P * L * U` where P
+  is a permutation matrix. 
+  
+  `L` and `U` are stored in the same matrix `LU` where 
+  the unit diagonal elements of L are not stored.
+  
+  `ipiv` contains the pivot indices such that row i of `A` 
+  was interchanged with row `ipiv(i)`.
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Please add a `:returns:` here and first state that this proc returns a tuple, where the first element is combination of `L` and `U` (as you already described in the paragraph above) and the second element is the pivot indices vector that contains ... (your explanation here). The most important thing being an explicit statement that this returns a tuple of size 2

File: modules/packages/LinearAlgebra.chpl [hunk 5]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+  
+*/
+
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: extra space

File: modules/packages/LinearAlgebra.chpl [hunk 6]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+    halt("Wrong rank for LU factorization");
+
+  if Adom.shape(1) != Adom.shape(2) then
+    halt("LU factorization only supports square matrices");
+
+  var (LU, ipiv, numSwap) = _lu(A);
+  return (LU,ipiv);
+}
+
+proc permute (ipiv, A: [?Adom] ?eltType, transpose=false) {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: I don't think you intend this proc to be user-facing, right? That's fine but then I think you need to do the following:
- make this a `private proc`
- add a `pragma "no doc"`
- a small free-form comment to describe what this function is doing (I guess this is optional as what it does should be clear in the linear algebra context)

Independent of this proc being not-user-facing, I think you should be explicit about `ipiv`s type, just to prevent mistakes.

File: modules/packages/LinearAlgebra.chpl [hunk 7]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+  const n = Adom.shape(1);
+  
+  var B: [Adom] eltType;
+  
+  if Adom.rank == 1 {
+    if transpose {
+      forall i in 1..n {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Note that this loop would likely perform better if written as:
```chapel
forall (i,pi) in zip(1..n, ipiv) {
  B[i] = A[pi];
}
```

... and similarly for the other loops that look like this one

File: modules/packages/LinearAlgebra.chpl [hunk 8]
```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
...
+
+  return (* reduce [i in Adom.dim(1)] LU[i,i]) * pdet;
+}
+
+/* Return the solution ``x`` to the linear system `` L * x = b `` 
+    where ``L`` is a lower triangular matrix. `unit_diag` will set the diagonal
+    elements to 1 within this procedure.
+*/
+proc solve_tril (const ref L: [?Ldom] ?eltType, const ref b: [?bdom] eltType, 
+                  unit_diag = true) {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: I don't quite understand the purpose of `unit_diag`. The explanation in the header comment, its use in the implementation, lack of it in `solve_triu` and the similar name but different behavior to scipy.linalg.solve_triangular's `unit_diagonal`. Is it supposed to be an optimization flag for cases where L is unit lower triangular?

File: test/library/packages/LinearAlgebra/performance/lu/lu-perf.chpl [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    writeln();
+  }
+
+  var A: [Space] eltType;
+
+  addDiag(A, 0, 2);
+  addDiag(A, 1, -1);
+  addDiag(A, -1, -1);
+  addDiag(A, 10, 0.5);
+  addDiag(A, -10, 0.5);
```
Comment (e-kayrakli) [inline test/library/packages/LinearAlgebra/performance/lu/lu-perf.chpl]: It is OK that this PR is dependent on your PR where you add these functions. But let's update these to setDiag as it is the interface that we settled on.

File: modules/packages/LinearAlgebra.chpl (hunk 3 above)
Comment (wongalvis14) [reply modules/packages/LinearAlgebra.chpl]: Yes, we could possibly be wasting 2 times more memory with the current implementation

File: modules/packages/LinearAlgebra.chpl (hunk 8 above)
Comment (wongalvis14) [reply modules/packages/LinearAlgebra.chpl]: It is intended to function like scipy.linalg.solve_triangular's unit_diagonal, I will change it accordingly.

**Comment by e-kayrakli** [comment]
@ben-albrecht -- where do you stand on adding the performance test here? Since this is a newly added function, I think we need a correctness test for sure. So:

- Do we add `.graph` to keep track of performance? And let this test be both performance and correctness test.
- Do we remove perf test related files to make it a correctness test?

**Comment by ben-albrecht** [comment]
> Do we add .graph to keep track of performance? And let this test be both performance and correctness test.
> Do we remove perf test related files to make it a correctness test?

I'm in favor of keeping the perf test as is, adding a `.graph` file and `GRAPHFILES` entry (under `# suite: Standard Library
`) for tracking performance of these routines, especially since they may receive further optimization attention in the future.

**Comment by ben-albrecht** [comment]
@wongalvis14  - I think it's been said offline, but we'll want to update the PR description to accurately reflect the changes made in this PR before merging.

**Comment by e-kayrakli** [comment]
> I'm in favor of keeping the perf test as is, adding a .graph file and GRAPHFILES entry (under # suite: Standard Library ) for tracking performance of these routines, especially since they may receive further optimization attention in the future.

There is an linearalgebra-perf.graph already in `test/library/packages/LinearAlgebra/` maybe we can lump this there.

**Comment by e-kayrakli** [comment]
@wongalvis14 -- You may have missed my previous comment. We could just use the existing linearalgebra-perf.graph to add these. That seems to be the right approach.

(If you were to add a new .graph file, you'd have to modify `test/GRAPHFILES` as well)

**Comment by e-kayrakli** [comment]
Also, the largest data size 10000 is too large, it times out in our nightly testing system (which has 5mins timeout). I think we should just remove it.

File: test/library/packages/LinearAlgebra/performance/lu/lu-perf.perfexecopts [hunk 10]
```diff
@@ -0,0 +1,2 @@
+--m=1000 --iters=100 -nl 1 #m1000
+--m=10000 --iters=10 -nl 1 #m10000
```
Comment (e-kayrakli) [inline test/library/packages/LinearAlgebra/performance/lu/lu-perf.perfexecopts]: Let's remove the larger size, and rename the file to have some "lu" in it somewhere. Current transpose tests also create files named "mXXX", and this one conflicts with those. (I think transpose's filenames should also change but not in this PR)
